File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
       int zcount = 0;
       vector<int>result = nums;
       for(auto it: nums){
           if(it == 0) zcount++;
       } 
       if(zcount > 1) return vector<int>(nums.size(),0);
       int prod = 1;
       for(auto it: nums){
           if(it) prod *= it;
       }  
       for(auto &it: result){
           if(zcount){
               it = it ? 0 : prod;
           }
           else
               it = prod / it;
       }
       return result;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        zero_pos = []
        prod = 1
        for ind, elt in enumerate(nums):
            if elt == 0:
                zero_pos.append(ind)
            else:
                prod = prod * elt
        if len(zero_pos) > 1:
            return [0] * len(nums)
        
        if len(zero_pos) == 1:
            result = [0] * len(nums)
            result[zero_pos[0]] = prod
            return result

        for i in range(len(nums)):
            nums[i] = int(prod / nums[i])
        return nums
        

